<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Ming-an Sun, Yejun Wang, Linlin Zou, Yanhua Li, Guoqiang Zhu, Todd Macfarlan" />

<meta name="date" content="2020-09-10" />

<title>ExprX-vignette</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">ExprX-vignette</h1>
<h4 class="author">Ming-an Sun, Yejun Wang, Linlin Zou, Yanhua Li, Guoqiang Zhu, Todd Macfarlan</h4>
<h4 class="date">2020-09-10</h4>



<div id="introduction" class="section level2">
<h2>1 Introduction</h2>
<p><strong>ExprX</strong> is an R package to streamline interspecies differential expression analysis. Taking TPM or FPKM/RPKM files for samples from different species as input, it provides functions to handle all the necessary steps, including data loading, ortholog matching, normalization, differential analysis and visualization.</p>
<p>ExprX is freely available at: <a href="https://github.com/mingansun/ExprX" class="uri">https://github.com/mingansun/ExprX</a>.</p>
</div>
<div id="how-to-install-exprx" class="section level2">
<h2>2 How to install ExprX</h2>
<p><strong>ExprX</strong> can be installed using the <strong>install_git</strong> function from devtools package. However, it depends on several other R packages, which should be installed first.</p>
<p>To install ExprX and its dependencies:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="co"># install several dependent packages from Bioconductor using BiocManager</span></a>
<a class="sourceLine" id="cb1-2" title="2"><span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</a>
<a class="sourceLine" id="cb1-3" title="3">BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="kw">c</span>(<span class="st">&quot;biomaRt&quot;</span>, <span class="st">&quot;edgeR&quot;</span>, <span class="st">&quot;RankProd&quot;</span>))</a>
<a class="sourceLine" id="cb1-4" title="4"></a>
<a class="sourceLine" id="cb1-5" title="5"><span class="co"># install ExprX from GitHub using devtools</span></a>
<a class="sourceLine" id="cb1-6" title="6"><span class="kw">install.packages</span>(<span class="st">&quot;devtools&quot;</span>)</a>
<a class="sourceLine" id="cb1-7" title="7">devtools<span class="op">::</span><span class="kw">install_git</span>(<span class="st">&quot;https://github.com/mingansun/ExprX&quot;</span>)</a></code></pre></div>
<p>To load the ExprX package:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1"><span class="kw">library</span>(ExprX)</a>
<a class="sourceLine" id="cb2-2" title="2"><span class="co">#&gt; Welcome to use ExprX!</span></a></code></pre></div>
<p>For more details about how to install and use <strong>ExprX</strong>, please refer to the website: <a href="https://github.com/mingansun/ExprX" class="uri">https://github.com/mingansun/ExprX</a></p>
</div>
<div id="apply-exprx-for-interspecies-differential-expression-analysis" class="section level2">
<h2>3 Apply ExprX for interspecies differential expression analysis</h2>
<p>Using RNA-Seq data for human and mouse brain, this vignette demonstrates how to detect differentially expressed genes between species using ExprX. All the involved steps are described below.</p>
<div id="generate-exprx-object-by-integrating-interspecies-expression-data" class="section level3">
<h3>3.1 Generate ExprX object by integrating interspecies expression data</h3>
<p>By parsing the meta table (as a data frame or CSV file) which contains information about expression data files (usually contain TPM, FPKM or RPKM values) for different species, the function <strong>make_ExprX_dataset</strong> can read these data files to create an object which contains the expression levels of the replicates of different species. The created ExprX object can also contain additional data such as orthologue pairs, normalized expression etc, and will be used by most of the subsequent analysis.</p>
<p>To use <strong>make_ExprX_dataset</strong> to read CSV file with meta data for expression data files and compared species (ie. human and mouse) to create an ExprX object:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1"><span class="co"># meta table file</span></a>
<a class="sourceLine" id="cb3-2" title="2">hs2mm.meta_file &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="kw">path.package</span>(<span class="st">&quot;ExprX&quot;</span>), <span class="st">&quot;/extdata/brain_metatable.csv&quot;</span>)</a>
<a class="sourceLine" id="cb3-3" title="3"></a>
<a class="sourceLine" id="cb3-4" title="4"><span class="co"># make ExprX object from meta table</span></a>
<a class="sourceLine" id="cb3-5" title="5">hs2mm.data  &lt;-<span class="st"> </span><span class="kw">make_ExprX_dataset</span>(</a>
<a class="sourceLine" id="cb3-6" title="6">  hs2mm.meta_file,</a>
<a class="sourceLine" id="cb3-7" title="7">  <span class="dt">data_dir =</span> <span class="kw">paste0</span>(<span class="kw">path.package</span>(<span class="st">&quot;ExprX&quot;</span>), <span class="st">&quot;/extdata&quot;</span>)</a>
<a class="sourceLine" id="cb3-8" title="8">  )</a>
<a class="sourceLine" id="cb3-9" title="9"><span class="co">#&gt; x is detected as a file name. Read to a data frame:</span></a>
<a class="sourceLine" id="cb3-10" title="10"><span class="co">#&gt; Species  FullName    AbbrName    IdColumn    ExprColumn  ExprType    RepIndex    File</span></a>
<a class="sourceLine" id="cb3-11" title="11"><span class="co">#&gt; human    Homo Sapiens    hsapiens    1   6   tpm 1   human_brain_1.genes.results</span></a>
<a class="sourceLine" id="cb3-12" title="12"><span class="co">#&gt; human    Homo Sapiens    hsapiens    1   6   tpm 2   human_brain_2.genes.results</span></a>
<a class="sourceLine" id="cb3-13" title="13"><span class="co">#&gt; human    Homo Sapiens    hsapiens    1   6   tpm 3   human_brain_3.genes.results</span></a>
<a class="sourceLine" id="cb3-14" title="14"><span class="co">#&gt; mouse    mmusculus   mmusculus   1   6   tpm 1   mouse_brain_1.genes.results</span></a>
<a class="sourceLine" id="cb3-15" title="15"><span class="co">#&gt; mouse    mmusculus   mmusculus   1   6   tpm 2   mouse_brain_2.genes.results</span></a>
<a class="sourceLine" id="cb3-16" title="16"><span class="co">#&gt; mouse    mmusculus   mmusculus   1   6   tpm 3   mouse_brain_3.genes.results</span></a>
<a class="sourceLine" id="cb3-17" title="17"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb3-18" title="18"><span class="co">#&gt; Read gene IDs for each species ...</span></a>
<a class="sourceLine" id="cb3-19" title="19"><span class="co">#&gt; human: 1042 genes.</span></a>
<a class="sourceLine" id="cb3-20" title="20"><span class="co">#&gt; mouse: 1037 genes.</span></a>
<a class="sourceLine" id="cb3-21" title="21"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb3-22" title="22"><span class="co">#&gt; Loading expression data for each species ...</span></a>
<a class="sourceLine" id="cb3-23" title="23"><span class="co">#&gt; human: 1042 genes from 3 files.</span></a>
<a class="sourceLine" id="cb3-24" title="24"><span class="co">#&gt; mouse: 1037 genes from 3 files.</span></a></code></pre></div>
</div>
<div id="determine-the-1-to-1-orthologs-among-compared-species" class="section level3">
<h3>3.2 Determine the 1-to-1 orthologs among compared species</h3>
<p>The 1-to-1 orthologs among species are constructed based on the homolog annotations from ENSEMBL database. Thus, only species available in in ENSEMBL database (about 200 as checked on 2020-4-7) can be used for analysis.</p>
<p>To use the <strong>list_species</strong> function to get the information (eg. species name, abbreviation) for all the species supported by ExprX:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1">sp.lst &lt;-<span class="st"> </span><span class="kw">list_species</span>()</a>
<a class="sourceLine" id="cb4-2" title="2"><span class="kw">head</span>(sp.lst)</a>
<a class="sourceLine" id="cb4-3" title="3"><span class="co">#&gt;         Dataset       Species                Version</span></a>
<a class="sourceLine" id="cb4-4" title="4"><span class="co">#&gt; 1   acalliptera Eastern happy             fAstCal1.2</span></a>
<a class="sourceLine" id="cb4-5" title="5"><span class="co">#&gt; 2 acarolinensis  Anole lizard              AnoCar2.0</span></a>
<a class="sourceLine" id="cb4-6" title="6"><span class="co">#&gt; 3  acchrysaetos  Golden eagle             bAquChr1.2</span></a>
<a class="sourceLine" id="cb4-7" title="7"><span class="co">#&gt; 4  acitrinellus Midas cichlid               Midas_v5</span></a>
<a class="sourceLine" id="cb4-8" title="8"><span class="co">#&gt; 5  amelanoleuca         Panda                ailMel1</span></a>
<a class="sourceLine" id="cb4-9" title="9"><span class="co">#&gt; 6    amexicanus Mexican tetra Astyanax_mexicanus-2.0</span></a></code></pre></div>
<div id="match-and-save-1-to-1-orthologs-among-species" class="section level4">
<h4>Match and save 1-to-1 orthologs among species</h4>
<p>The <strong>ortholog_match</strong> function invokes <strong>biomaRt</strong> package to retrieve homolog annotation from ENSEMBL database, then matches 1-to-1 orthologs by reciprocal comparison. This step usually takes a few minutes - depending on the network speed). To speed up, the obtained ortholog data can be stored on hard disk with <strong>saveRDS</strong> for later use.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1"><span class="co"># Match 1:1 orthologs between human and mouse</span></a>
<a class="sourceLine" id="cb5-2" title="2">hs2mm.orth &lt;-<span class="st"> </span><span class="kw">ortholog_match</span>(<span class="st">&quot;human&quot;</span>, <span class="st">&quot;mouse&quot;</span>)</a>
<a class="sourceLine" id="cb5-3" title="3"><span class="co">#&gt; Ensembl site unresponsive, trying useast mirror</span></a>
<a class="sourceLine" id="cb5-4" title="4"><span class="co">#&gt; Ensembl site unresponsive, trying uswest mirror</span></a>
<a class="sourceLine" id="cb5-5" title="5"><span class="co">#&gt; Get species name abbreviation to be used by biomaRt ...</span></a>
<a class="sourceLine" id="cb5-6" title="6"><span class="co">#&gt; human =&gt; hsapiens</span></a>
<a class="sourceLine" id="cb5-7" title="7"><span class="co">#&gt; mouse =&gt; mmusculus</span></a>
<a class="sourceLine" id="cb5-8" title="8"><span class="co">#&gt; Attempting web service request:</span></a>
<a class="sourceLine" id="cb5-9" title="9"><span class="co">#&gt; http://www.ensembl.org:80/biomart/martservice?type=version&amp;requestid=biomaRt&amp;mart=ENSEMBL_MART_ENSEMBL</span></a>
<a class="sourceLine" id="cb5-10" title="10"><span class="co">#&gt;    V1</span></a>
<a class="sourceLine" id="cb5-11" title="11"><span class="co">#&gt; 1 0.7</span></a>
<a class="sourceLine" id="cb5-12" title="12"><span class="co">#&gt; BioMartServer running BioMart version: 0.7</span></a>
<a class="sourceLine" id="cb5-13" title="13"><span class="co">#&gt; Mart virtual schema: default</span></a>
<a class="sourceLine" id="cb5-14" title="14"><span class="co">#&gt; Mart host: http://www.ensembl.org:80/biomart/martservice</span></a>
<a class="sourceLine" id="cb5-15" title="15"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb5-16" title="16"><span class="co">#&gt; Connect to biomaRt for each species ...</span></a>
<a class="sourceLine" id="cb5-17" title="17"><span class="co">#&gt; human =&gt; hsapiens</span></a>
<a class="sourceLine" id="cb5-18" title="18"><span class="co">#&gt; mouse =&gt; mmusculus</span></a>
<a class="sourceLine" id="cb5-19" title="19"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb5-20" title="20"><span class="co">#&gt; Determine the homolog table to be retrieved ...</span></a>
<a class="sourceLine" id="cb5-21" title="21"><span class="co">#&gt; human =&gt; mouse:  mmusculus_homolog_ensembl_gene</span></a>
<a class="sourceLine" id="cb5-22" title="22"><span class="co">#&gt; mouse =&gt; human:  hsapiens_homolog_ensembl_gene</span></a>
<a class="sourceLine" id="cb5-23" title="23"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb5-24" title="24"><span class="co">#&gt; Retrieving homolog annotation for human ...</span></a>
<a class="sourceLine" id="cb5-25" title="25"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb5-26" title="26"><span class="co">#&gt; Retrieving homolog annotation for mouse ...</span></a>
<a class="sourceLine" id="cb5-27" title="27"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb5-28" title="28"><span class="co">#&gt; Number of genes with only 1 match in other species:</span></a>
<a class="sourceLine" id="cb5-29" title="29"><span class="co">#&gt; human:   18330</span></a>
<a class="sourceLine" id="cb5-30" title="30"><span class="co">#&gt; mouse:   18873</span></a>
<a class="sourceLine" id="cb5-31" title="31"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb5-32" title="32"><span class="co">#&gt; Number of 1-to-1 ortholog pairs before filtering:    16758</span></a>
<a class="sourceLine" id="cb5-33" title="33"><span class="co">#&gt; Number of 1-to-1 ortholog pairs after  filtering:    16758</span></a>
<a class="sourceLine" id="cb5-34" title="34"></a>
<a class="sourceLine" id="cb5-35" title="35"><span class="co"># Save the ortholog results on hard disk for later use</span></a>
<a class="sourceLine" id="cb5-36" title="36"><span class="kw">saveRDS</span>(hs2mm.orth, <span class="st">&quot;hs2mm.orth.rds&quot;</span>)</a></code></pre></div>
</div>
<div id="load-previously-saved-ortholog-matching-results-from-hard-disk" class="section level4">
<h4>Load previously saved ortholog matching results from hard disk</h4>
<p>Below shows how to used <strong>readRDS</strong> to load ortholog matching result that is previously saved on hard disk. Alternatively, the ortholog matching result can be generated with <strong>ortholog_match</strong> as demonstrated above.</p>
<p>The ortholog matching result includes information for each involved species, such as GeneID, GeneName, Chrom, GeneType and so on.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" title="1"><span class="co"># Load ortholog result with readRDS</span></a>
<a class="sourceLine" id="cb6-2" title="2">hs2mm.orth &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="kw">paste0</span>(<span class="kw">path.package</span>(<span class="st">&quot;ExprX&quot;</span>), <span class="st">&quot;/data/hs2mm.orth.rds&quot;</span>))</a>
<a class="sourceLine" id="cb6-3" title="3"></a>
<a class="sourceLine" id="cb6-4" title="4"><span class="co"># View the structure of hs2mm.orth</span></a>
<a class="sourceLine" id="cb6-5" title="5"><span class="kw">str</span>(hs2mm.orth)</a>
<a class="sourceLine" id="cb6-6" title="6"><span class="co">#&gt; List of 2</span></a>
<a class="sourceLine" id="cb6-7" title="7"><span class="co">#&gt;  $ human:&#39;data.frame&#39;:   16536 obs. of  4 variables:</span></a>
<a class="sourceLine" id="cb6-8" title="8"><span class="co">#&gt;   ..$ GeneID  : chr [1:16536] &quot;ENSG00000198695&quot; &quot;ENSG00000198712&quot; &quot;ENSG00000198727&quot; &quot;ENSG00000198763&quot; ...</span></a>
<a class="sourceLine" id="cb6-9" title="9"><span class="co">#&gt;   ..$ GeneName: chr [1:16536] &quot;MT-ND6&quot; &quot;MT-CO2&quot; &quot;MT-CYB&quot; &quot;MT-ND2&quot; ...</span></a>
<a class="sourceLine" id="cb6-10" title="10"><span class="co">#&gt;   ..$ Chrom   : chr [1:16536] &quot;MT&quot; &quot;MT&quot; &quot;MT&quot; &quot;MT&quot; ...</span></a>
<a class="sourceLine" id="cb6-11" title="11"><span class="co">#&gt;   ..$ GeneType: chr [1:16536] &quot;protein_coding&quot; &quot;protein_coding&quot; &quot;protein_coding&quot; &quot;protein_coding&quot; ...</span></a>
<a class="sourceLine" id="cb6-12" title="12"><span class="co">#&gt;  $ mouse:&#39;data.frame&#39;:   16536 obs. of  4 variables:</span></a>
<a class="sourceLine" id="cb6-13" title="13"><span class="co">#&gt;   ..$ GeneID  : chr [1:16536] &quot;ENSMUSG00000064368&quot; &quot;ENSMUSG00000064354&quot; &quot;ENSMUSG00000064370&quot; &quot;ENSMUSG00000064345&quot; ...</span></a>
<a class="sourceLine" id="cb6-14" title="14"><span class="co">#&gt;   ..$ GeneName: chr [1:16536] &quot;mt-Nd6&quot; &quot;mt-Co2&quot; &quot;mt-Cytb&quot; &quot;mt-Nd2&quot; ...</span></a>
<a class="sourceLine" id="cb6-15" title="15"><span class="co">#&gt;   ..$ Chrom   : chr [1:16536] &quot;MT&quot; &quot;MT&quot; &quot;MT&quot; &quot;MT&quot; ...</span></a>
<a class="sourceLine" id="cb6-16" title="16"><span class="co">#&gt;   ..$ GeneType: chr [1:16536] &quot;protein_coding&quot; &quot;protein_coding&quot; &quot;protein_coding&quot; &quot;protein_coding&quot; ...</span></a>
<a class="sourceLine" id="cb6-17" title="17"><span class="co">#&gt;  - attr(*, &quot;Species&quot;)= chr [1:2] &quot;human&quot; &quot;mouse&quot;</span></a>
<a class="sourceLine" id="cb6-18" title="18"><span class="co">#&gt;  - attr(*, &quot;SpeciesAbbr&quot;)= chr [1:2] &quot;hsapiens&quot; &quot;mmusculus&quot;</span></a>
<a class="sourceLine" id="cb6-19" title="19"><span class="co">#&gt;  - attr(*, &quot;SpeciesFull&quot;)= chr [1:2] &quot;Homo Sapiens&quot; &quot;mmusculus&quot;</span></a></code></pre></div>
<p>To summarize the ortholog results by genetype (eg. protein_coding, miRNA, lncRNA etc):</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1">hs2mm.orth.genetype &lt;-<span class="st"> </span><span class="kw">summarize_ortholog_gene</span>(hs2mm.orth, <span class="dt">group =</span> <span class="st">&quot;genetype&quot;</span>)</a>
<a class="sourceLine" id="cb7-2" title="2"><span class="kw">head</span>(hs2mm.orth.genetype)</a>
<a class="sourceLine" id="cb7-3" title="3"><span class="co">#&gt;           human mouse</span></a>
<a class="sourceLine" id="cb7-4" title="4"><span class="co">#&gt; IG_C_gene     3     3</span></a>
<a class="sourceLine" id="cb7-5" title="5"><span class="co">#&gt; IG_V_gene     2     2</span></a>
<a class="sourceLine" id="cb7-6" title="6"><span class="co">#&gt; TR_C_gene     1     1</span></a>
<a class="sourceLine" id="cb7-7" title="7"><span class="co">#&gt; TR_V_gene    16    16</span></a>
<a class="sourceLine" id="cb7-8" title="8"><span class="co">#&gt; lncRNA        1     0</span></a>
<a class="sourceLine" id="cb7-9" title="9"><span class="co">#&gt; miRNA       218   218</span></a></code></pre></div>
<p>To summarize the ortholog matching results by chromosome:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1">hs2mm.orth.chrom &lt;-<span class="st"> </span><span class="kw">summarize_ortholog_gene</span>(hs2mm.orth, <span class="dt">group =</span> <span class="st">&quot;chrom&quot;</span>)</a>
<a class="sourceLine" id="cb8-2" title="2"><span class="kw">head</span>(hs2mm.orth.chrom)</a>
<a class="sourceLine" id="cb8-3" title="3"><span class="co">#&gt;    human mouse</span></a>
<a class="sourceLine" id="cb8-4" title="4"><span class="co">#&gt; 1   1727  1050</span></a>
<a class="sourceLine" id="cb8-5" title="5"><span class="co">#&gt; 10   636   779</span></a>
<a class="sourceLine" id="cb8-6" title="6"><span class="co">#&gt; 11  1048  1342</span></a>
<a class="sourceLine" id="cb8-7" title="7"><span class="co">#&gt; 12   889   565</span></a>
<a class="sourceLine" id="cb8-8" title="8"><span class="co">#&gt; 13   286   546</span></a>
<a class="sourceLine" id="cb8-9" title="9"><span class="co">#&gt; 14   572   588</span></a></code></pre></div>
</div>
<div id="filter-1-to-1-orthologs-to-exclude-specific-groups-of-genes" class="section level4">
<h4>Filter 1-to-1 orthologs to exclude specific groups of genes</h4>
<p>In many cases, specific groups of genes (eg. pseudogenes or genes from sex chromosomes) are undesirable for gene expression comparison. The function <strong>ortholog_filter</strong> enables the filtering of ortholog pairs based on gene type, chromosome, or provided gene list, as demonstrated below.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1"><span class="co"># Filter orthologs by excluding genes from chromosomes X, Y and MT, and only keep</span></a>
<a class="sourceLine" id="cb9-2" title="2"><span class="co"># protein_coding genes</span></a>
<a class="sourceLine" id="cb9-3" title="3">hs2mm.orth.flt &lt;-<span class="st"> </span><span class="kw">ortholog_filter</span>(</a>
<a class="sourceLine" id="cb9-4" title="4">  hs2mm.orth,</a>
<a class="sourceLine" id="cb9-5" title="5">  <span class="dt">genetype_include =</span> <span class="st">&quot;protein_coding&quot;</span>,</a>
<a class="sourceLine" id="cb9-6" title="6">  <span class="dt">chrom_exclude =</span> <span class="kw">c</span>(<span class="st">&quot;X&quot;</span>, <span class="st">&quot;Y&quot;</span>, <span class="st">&quot;MT&quot;</span>)</a>
<a class="sourceLine" id="cb9-7" title="7">)</a>
<a class="sourceLine" id="cb9-8" title="8"><span class="co">#&gt; Check data for human</span></a>
<a class="sourceLine" id="cb9-9" title="9"><span class="co">#&gt; genetype_include matches:    15849</span></a>
<a class="sourceLine" id="cb9-10" title="10"><span class="co">#&gt; chrom_exclude    matches:    595</span></a>
<a class="sourceLine" id="cb9-11" title="11"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb9-12" title="12"><span class="co">#&gt; Check data for mouse</span></a>
<a class="sourceLine" id="cb9-13" title="13"><span class="co">#&gt; genetype_include matches:    15855</span></a>
<a class="sourceLine" id="cb9-14" title="14"><span class="co">#&gt; chrom_exclude    matches:    590</span></a>
<a class="sourceLine" id="cb9-15" title="15"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb9-16" title="16"><span class="co">#&gt; Original gene number: 16536</span></a>
<a class="sourceLine" id="cb9-17" title="17"><span class="co">#&gt; Filtered gene number: 15275</span></a></code></pre></div>
<p>To check the number of 1:1 orthologs before and after filtering:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1"><span class="co"># Before filtering</span></a>
<a class="sourceLine" id="cb10-2" title="2"><span class="kw">dim</span>(hs2mm.orth[[<span class="dv">1</span>]])</a>
<a class="sourceLine" id="cb10-3" title="3"><span class="co">#&gt; [1] 16536     4</span></a>
<a class="sourceLine" id="cb10-4" title="4"></a>
<a class="sourceLine" id="cb10-5" title="5"><span class="co"># After filtering</span></a>
<a class="sourceLine" id="cb10-6" title="6"><span class="kw">dim</span>(hs2mm.orth.flt[[<span class="dv">1</span>]])</a>
<a class="sourceLine" id="cb10-7" title="7"><span class="co">#&gt; [1] 15275     4</span></a></code></pre></div>
</div>
</div>
<div id="integrate-the-1-to-1-ortholog-matching-result-to-exprx-object" class="section level3">
<h3>3.3 Integrate the 1-to-1 ortholog matching result to ExprX object</h3>
<p>Take the original ExprX object generated with <strong>make_ExprX_dataset</strong> and the ortholog matching data generated using <strong>ortholog_match</strong> as input, the function <strong>ortholog_expression_merge</strong> integrates together the expression data for all 1:1 orthologs among compared species. The integrated data will be appended to the original ExprX object and returned as an updated object. To be noted, for orthologs that donâ€™t have matched expression data, they will be excluded from the returned data.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" title="1"><span class="co"># Merge data</span></a>
<a class="sourceLine" id="cb11-2" title="2">hs2mm.data &lt;-<span class="st"> </span><span class="kw">ortholog_expression_merge</span>(</a>
<a class="sourceLine" id="cb11-3" title="3">  <span class="dt">expr_data =</span> hs2mm.data, <span class="dt">orth_data =</span> hs2mm.orth.flt</a>
<a class="sourceLine" id="cb11-4" title="4">)</a>
<a class="sourceLine" id="cb11-5" title="5"><span class="co">#&gt; Number of ortholog pairs absent from expr_data</span></a>
<a class="sourceLine" id="cb11-6" title="6"><span class="co">#&gt; human:   14233</span></a>
<a class="sourceLine" id="cb11-7" title="7"><span class="co">#&gt; mouse:   14238</span></a>
<a class="sourceLine" id="cb11-8" title="8"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb11-9" title="9"><span class="co">#&gt; Original  ortholog number: 15275</span></a>
<a class="sourceLine" id="cb11-10" title="10"><span class="co">#&gt; Discarded ortholog number: 14375</span></a>
<a class="sourceLine" id="cb11-11" title="11"><span class="co">#&gt; Resulted  ortholog number: 900</span></a></code></pre></div>
</div>
<div id="integrate-the-normalized-expression-data-for-1-to-1-orthologs-to-exprx-object" class="section level3">
<h3>3.4 Integrate the normalized expression data for 1-to-1 orthologs to ExprX object</h3>
<p>Normalization of the expression data for 1:1 orthologs among species can be performed by using the <strong>ortholog_expression_normalize</strong> function. Different normalization approaches are supported, including TMM, TMMwsp, RLE, upperquartile and quantile. The normalized data matrix will be appended to the original ExprX object and returned as the updated ExprX object.</p>
<p>To normalize the expression data of 1:1 orthologs among samples using the TMM approach:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" title="1"><span class="co"># Load required package</span></a>
<a class="sourceLine" id="cb12-2" title="2"><span class="kw">library</span>(edgeR)</a>
<a class="sourceLine" id="cb12-3" title="3"><span class="co">#&gt; Loading required package: limma</span></a>
<a class="sourceLine" id="cb12-4" title="4"></a>
<a class="sourceLine" id="cb12-5" title="5"><span class="co"># Perform data normalization</span></a>
<a class="sourceLine" id="cb12-6" title="6">hs2mm.data &lt;-<span class="st"> </span><span class="kw">ortholog_expression_normalize</span>(</a>
<a class="sourceLine" id="cb12-7" title="7">  <span class="dt">expr_data =</span> hs2mm.data, <span class="dt">method =</span> <span class="st">&quot;TMM&quot;</span></a>
<a class="sourceLine" id="cb12-8" title="8">)</a></code></pre></div>
</div>
<div id="perform-interspecies-differential-expression-analysis" class="section level3">
<h3>3.5 Perform interspecies differential expression analysis</h3>
<p>Differential expression analysis of 1:1 orthologs between species can be performed using the <strong>ortholog_expression_compare</strong> function. Expression data of 1:1 orthologs after normalization are used for differential expression analysis. Statistics such as average expression level, log2foldChange and p-values are calculated and returned as a dataframe.</p>
<p>Below shows how to perform interspecies differential analysis for 1:1 orthologs using RankProd approach. To be noted, the demo dataset is only for less than 1000 genes, thus only a couple of genes are called as differentlly expressed. If use full dataset, the number of called differential genes can be as many as several hundreds.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" title="1"><span class="co"># Load required package</span></a>
<a class="sourceLine" id="cb13-2" title="2"><span class="kw">library</span>(RankProd)</a>
<a class="sourceLine" id="cb13-3" title="3"><span class="co">#&gt; Loading required package: Rmpfr</span></a>
<a class="sourceLine" id="cb13-4" title="4"><span class="co">#&gt; Loading required package: gmp</span></a>
<a class="sourceLine" id="cb13-5" title="5"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb13-6" title="6"><span class="co">#&gt; Attaching package: &#39;gmp&#39;</span></a>
<a class="sourceLine" id="cb13-7" title="7"><span class="co">#&gt; The following objects are masked from &#39;package:base&#39;:</span></a>
<a class="sourceLine" id="cb13-8" title="8"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb13-9" title="9"><span class="co">#&gt;     %*%, apply, crossprod, matrix, tcrossprod</span></a>
<a class="sourceLine" id="cb13-10" title="10"><span class="co">#&gt; C code of R package &#39;Rmpfr&#39;: GMP using 64 bits per limb</span></a>
<a class="sourceLine" id="cb13-11" title="11"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb13-12" title="12"><span class="co">#&gt; Attaching package: &#39;Rmpfr&#39;</span></a>
<a class="sourceLine" id="cb13-13" title="13"><span class="co">#&gt; The following object is masked from &#39;package:gmp&#39;:</span></a>
<a class="sourceLine" id="cb13-14" title="14"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb13-15" title="15"><span class="co">#&gt;     outer</span></a>
<a class="sourceLine" id="cb13-16" title="16"><span class="co">#&gt; The following objects are masked from &#39;package:stats&#39;:</span></a>
<a class="sourceLine" id="cb13-17" title="17"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb13-18" title="18"><span class="co">#&gt;     dbinom, dgamma, dnorm, dpois, pnorm</span></a>
<a class="sourceLine" id="cb13-19" title="19"><span class="co">#&gt; The following objects are masked from &#39;package:base&#39;:</span></a>
<a class="sourceLine" id="cb13-20" title="20"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb13-21" title="21"><span class="co">#&gt;     cbind, pmax, pmin, rbind</span></a>
<a class="sourceLine" id="cb13-22" title="22"></a>
<a class="sourceLine" id="cb13-23" title="23"><span class="co"># Perform differential analysis and then sort by p-value</span></a>
<a class="sourceLine" id="cb13-24" title="24">hs2mm.deg &lt;-<span class="st"> </span><span class="kw">ortholog_expression_compare</span>(</a>
<a class="sourceLine" id="cb13-25" title="25">  hs2mm.data, <span class="dt">method =</span> <span class="st">&quot;RankProd&quot;</span>, <span class="dt">p_adjust =</span> <span class="st">&quot;fdr&quot;</span></a>
<a class="sourceLine" id="cb13-26" title="26">  )</a>
<a class="sourceLine" id="cb13-27" title="27"><span class="co">#&gt; Rank Product analysis for unpaired case </span></a>
<a class="sourceLine" id="cb13-28" title="28"><span class="co">#&gt;  </span></a>
<a class="sourceLine" id="cb13-29" title="29"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb13-30" title="30"><span class="co">#&gt;  done</span></a>
<a class="sourceLine" id="cb13-31" title="31">hs2mm.deg &lt;-<span class="st"> </span>hs2mm.deg[<span class="kw">order</span>(hs2mm.deg<span class="op">$</span>P_value),]</a>
<a class="sourceLine" id="cb13-32" title="32"></a>
<a class="sourceLine" id="cb13-33" title="33"><span class="co"># Check what the differential analysis result looks like</span></a>
<a class="sourceLine" id="cb13-34" title="34"><span class="kw">head</span>(hs2mm.deg)</a>
<a class="sourceLine" id="cb13-35" title="35"><span class="co">#&gt;        GeneID_human       GeneID_mouse GeneName_human GeneName_mouse</span></a>
<a class="sourceLine" id="cb13-36" title="36"><span class="co">#&gt; 346 ENSG00000089220 ENSMUSG00000047104          PEBP1           Pbp2</span></a>
<a class="sourceLine" id="cb13-37" title="37"><span class="co">#&gt; 233 ENSG00000131095 ENSMUSG00000020932           GFAP           Gfap</span></a>
<a class="sourceLine" id="cb13-38" title="38"><span class="co">#&gt; 198 ENSG00000132341 ENSMUSG00000028287            RAN  1700009N14Rik</span></a>
<a class="sourceLine" id="cb13-39" title="39"><span class="co">#&gt; 733 ENSG00000185559 ENSMUSG00000040856           DLK1           Dlk1</span></a>
<a class="sourceLine" id="cb13-40" title="40"><span class="co">#&gt; 533 ENSG00000129538 ENSMUSG00000035896         RNASE1         Rnase1</span></a>
<a class="sourceLine" id="cb13-41" title="41"><span class="co">#&gt; 222 ENSG00000171794 ENSMUSG00000051396           UTF1        Gm45902</span></a>
<a class="sourceLine" id="cb13-42" title="42"><span class="co">#&gt;     Expression_human Expression_mouse Expression_average log2foldChange</span></a>
<a class="sourceLine" id="cb13-43" title="43"><span class="co">#&gt; 346     1.160654e+04        1.0046551          5803.7744      12.913280</span></a>
<a class="sourceLine" id="cb13-44" title="44"><span class="co">#&gt; 233     4.798557e+04      359.1007821         24172.3349       7.060076</span></a>
<a class="sourceLine" id="cb13-45" title="45"><span class="co">#&gt; 198     4.448541e+03        0.7653213          2224.6534      11.779775</span></a>
<a class="sourceLine" id="cb13-46" title="46"><span class="co">#&gt; 733     8.359417e-01     2378.4807928          1189.6584     -10.798271</span></a>
<a class="sourceLine" id="cb13-47" title="47"><span class="co">#&gt; 533     5.322110e+03        3.8217766          2662.9657      10.266294</span></a>
<a class="sourceLine" id="cb13-48" title="48"><span class="co">#&gt; 222     5.568115e-01      617.4629126           309.0099      -9.191658</span></a>
<a class="sourceLine" id="cb13-49" title="49"><span class="co">#&gt;          P_value</span></a>
<a class="sourceLine" id="cb13-50" title="50"><span class="co">#&gt; 346 1.504507e-05</span></a>
<a class="sourceLine" id="cb13-51" title="51"><span class="co">#&gt; 233 9.797977e-05</span></a>
<a class="sourceLine" id="cb13-52" title="52"><span class="co">#&gt; 198 1.016858e-04</span></a>
<a class="sourceLine" id="cb13-53" title="53"><span class="co">#&gt; 733 2.257631e-04</span></a>
<a class="sourceLine" id="cb13-54" title="54"><span class="co">#&gt; 533 4.537869e-04</span></a>
<a class="sourceLine" id="cb13-55" title="55"><span class="co">#&gt; 222 9.593216e-04</span></a></code></pre></div>
<p>To determine differential genes (based on cufoff of p-value and log2foldChange) which can be saved for downstream analysis:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" title="1"><span class="co"># Determine significant differential genes (human&gt;mouse) based on p-values and log2foldChange</span></a>
<a class="sourceLine" id="cb14-2" title="2">hs2mm.hsHigh &lt;-<span class="st"> </span><span class="kw">subset</span>(hs2mm.deg, <span class="dt">subset =</span> log2foldChange <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>  <span class="op">&amp;</span><span class="st"> </span>P_value <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.05</span>)</a>
<a class="sourceLine" id="cb14-3" title="3"><span class="kw">head</span>(hs2mm.hsHigh)</a>
<a class="sourceLine" id="cb14-4" title="4"><span class="co">#&gt;        GeneID_human       GeneID_mouse GeneName_human GeneName_mouse</span></a>
<a class="sourceLine" id="cb14-5" title="5"><span class="co">#&gt; 346 ENSG00000089220 ENSMUSG00000047104          PEBP1           Pbp2</span></a>
<a class="sourceLine" id="cb14-6" title="6"><span class="co">#&gt; 233 ENSG00000131095 ENSMUSG00000020932           GFAP           Gfap</span></a>
<a class="sourceLine" id="cb14-7" title="7"><span class="co">#&gt; 198 ENSG00000132341 ENSMUSG00000028287            RAN  1700009N14Rik</span></a>
<a class="sourceLine" id="cb14-8" title="8"><span class="co">#&gt; 533 ENSG00000129538 ENSMUSG00000035896         RNASE1         Rnase1</span></a>
<a class="sourceLine" id="cb14-9" title="9"><span class="co">#&gt; 504 ENSG00000145945 ENSMUSG00000038246         FAM50B         Fam50b</span></a>
<a class="sourceLine" id="cb14-10" title="10"><span class="co">#&gt; 190 ENSG00000204655 ENSMUSG00000076439            MOG            Mog</span></a>
<a class="sourceLine" id="cb14-11" title="11"><span class="co">#&gt;     Expression_human Expression_mouse Expression_average log2foldChange</span></a>
<a class="sourceLine" id="cb14-12" title="12"><span class="co">#&gt; 346       11606.5442        1.0046551          5803.7744      12.913280</span></a>
<a class="sourceLine" id="cb14-13" title="13"><span class="co">#&gt; 233       47985.5690      359.1007821         24172.3349       7.060076</span></a>
<a class="sourceLine" id="cb14-14" title="14"><span class="co">#&gt; 198        4448.5415        0.7653213          2224.6534      11.779775</span></a>
<a class="sourceLine" id="cb14-15" title="15"><span class="co">#&gt; 533        5322.1096        3.8217766          2662.9657      10.266294</span></a>
<a class="sourceLine" id="cb14-16" title="16"><span class="co">#&gt; 504         312.7094        0.2713377           156.4903       8.665549</span></a>
<a class="sourceLine" id="cb14-17" title="17"><span class="co">#&gt; 190        3020.8812      399.3283886          1710.1048       2.917755</span></a>
<a class="sourceLine" id="cb14-18" title="18"><span class="co">#&gt;          P_value</span></a>
<a class="sourceLine" id="cb14-19" title="19"><span class="co">#&gt; 346 1.504507e-05</span></a>
<a class="sourceLine" id="cb14-20" title="20"><span class="co">#&gt; 233 9.797977e-05</span></a>
<a class="sourceLine" id="cb14-21" title="21"><span class="co">#&gt; 198 1.016858e-04</span></a>
<a class="sourceLine" id="cb14-22" title="22"><span class="co">#&gt; 533 4.537869e-04</span></a>
<a class="sourceLine" id="cb14-23" title="23"><span class="co">#&gt; 504 2.639805e-03</span></a>
<a class="sourceLine" id="cb14-24" title="24"><span class="co">#&gt; 190 1.770427e-02</span></a>
<a class="sourceLine" id="cb14-25" title="25"></a>
<a class="sourceLine" id="cb14-26" title="26"><span class="co"># Determine significant differential genes (human&lt;mouse) based on p-values and log2foldChange</span></a>
<a class="sourceLine" id="cb14-27" title="27">hs2mm.mmHigh &lt;-<span class="st"> </span><span class="kw">subset</span>(hs2mm.deg, <span class="dt">subset =</span> log2foldChange <span class="op">&lt;</span><span class="st"> </span><span class="dv">-1</span> <span class="op">&amp;</span><span class="st"> </span>P_value <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.05</span>)</a>
<a class="sourceLine" id="cb14-28" title="28"><span class="kw">head</span>(hs2mm.mmHigh)</a>
<a class="sourceLine" id="cb14-29" title="29"><span class="co">#&gt;        GeneID_human       GeneID_mouse GeneName_human GeneName_mouse</span></a>
<a class="sourceLine" id="cb14-30" title="30"><span class="co">#&gt; 733 ENSG00000185559 ENSMUSG00000040856           DLK1           Dlk1</span></a>
<a class="sourceLine" id="cb14-31" title="31"><span class="co">#&gt; 222 ENSG00000171794 ENSMUSG00000051396           UTF1        Gm45902</span></a>
<a class="sourceLine" id="cb14-32" title="32"><span class="co">#&gt; 389 ENSG00000170236 ENSMUSG00000027364          USP50          Usp50</span></a>
<a class="sourceLine" id="cb14-33" title="33"><span class="co">#&gt; 618 ENSG00000008196 ENSMUSG00000025927         TFAP2B         Tfap2b</span></a>
<a class="sourceLine" id="cb14-34" title="34"><span class="co">#&gt; 385 ENSG00000151079 ENSMUSG00000038077          KCNA6          Kcna6</span></a>
<a class="sourceLine" id="cb14-35" title="35"><span class="co">#&gt; 507 ENSG00000170561 ENSMUSG00000001504           IRX2           Irx2</span></a>
<a class="sourceLine" id="cb14-36" title="36"><span class="co">#&gt;     Expression_human Expression_mouse Expression_average log2foldChange</span></a>
<a class="sourceLine" id="cb14-37" title="37"><span class="co">#&gt; 733       0.83594167        2378.4808          1189.6584     -10.798271</span></a>
<a class="sourceLine" id="cb14-38" title="38"><span class="co">#&gt; 222       0.55681154         617.4629           309.0099      -9.191658</span></a>
<a class="sourceLine" id="cb14-39" title="39"><span class="co">#&gt; 389       3.63377004        1604.9029           804.2683      -8.601262</span></a>
<a class="sourceLine" id="cb14-40" title="40"><span class="co">#&gt; 618       0.08851788         631.8555           315.9720     -10.069434</span></a>
<a class="sourceLine" id="cb14-41" title="41"><span class="co">#&gt; 385       0.00000000         296.3993           148.1996      -9.213830</span></a>
<a class="sourceLine" id="cb14-42" title="42"><span class="co">#&gt; 507       1.02556448         608.3116           304.6686      -8.640509</span></a>
<a class="sourceLine" id="cb14-43" title="43"><span class="co">#&gt;          P_value</span></a>
<a class="sourceLine" id="cb14-44" title="44"><span class="co">#&gt; 733 0.0002257631</span></a>
<a class="sourceLine" id="cb14-45" title="45"><span class="co">#&gt; 222 0.0009593216</span></a>
<a class="sourceLine" id="cb14-46" title="46"><span class="co">#&gt; 389 0.0010448787</span></a>
<a class="sourceLine" id="cb14-47" title="47"><span class="co">#&gt; 618 0.0011482998</span></a>
<a class="sourceLine" id="cb14-48" title="48"><span class="co">#&gt; 385 0.0021331634</span></a>
<a class="sourceLine" id="cb14-49" title="49"><span class="co">#&gt; 507 0.0021331634</span></a></code></pre></div>
</div>
<div id="visualize-differential-expression-analysis-result" class="section level3">
<h3>3.6 Visualize differential expression analysis result</h3>
<p>The differential expression can be visualized as MA-plot or Volcano-plot, with differential genes highlighted by color. Below shows how to use the <strong>ortholog_expression_plot</strong> function to generate MA-plot and Volcano-plot, respectively.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" title="1"><span class="co"># Generate MA-plot</span></a>
<a class="sourceLine" id="cb15-2" title="2"><span class="kw">ortholog_expression_plot</span>(</a>
<a class="sourceLine" id="cb15-3" title="3">  hs2mm.deg, <span class="st">&quot;MA&quot;</span>,</a>
<a class="sourceLine" id="cb15-4" title="4">  <span class="dt">main =</span> <span class="st">&quot;Brain&quot;</span>, <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">120000</span>), <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">18</span>,<span class="dv">18</span>)</a>
<a class="sourceLine" id="cb15-5" title="5">  )</a></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAYAAAAGACAMAAACTGUWNAAAApVBMVEUAAAAAACsAAFUAK1UAK4AAVYAAVaorAAArACsrAFUrKwArK1UrVYArgNRVAABVACtVAFVVKwBVVStVVVVVgIBVqqpVqtRVqv+AKwCAVSuAgFWAgKqAqoCAqtSA1NSA1P+qVQCqVSuqgFWqqlWq1ICq1P+q/6qq//++vr7UgCvUqoDU1KrU1NTU/9TU////AAD/qlX/wMv/1ID/1Kr//6r//9T///+eX/NWAAAACXBIWXMAAA7DAAAOwwHHb6hkAAANJElEQVR4nO2dCZujtgGGmVnXnjS1N1c7bpp23GN2TdJ6qGv7//+0ogt8AbrwB+h7n2djMBpB9FoXSCI7ESgZ+gJShwLAUAAYCgBDAWAoAAwFgKEAMBQAhgLAUAAYCgBDAWAoAAwFgKEAMBQAhgLAUAAYCgBDAWAoAAwFgKEAMBQAhgLAUAAYCgBDAWAoAAwFgKEAMGMWkGeKpz+83z1+XGez3YOvyZkpCMiy56/3jlNAz9QCsiX6WrwZtwD1y98vsrnYe/rl++z5/fTr96WRb99MDii//9vfs+zTG/py7zMFAb8uRA5Q+WG2K3TF8FYJqL4YIuMWYBBFfbk3ez/9u0z13+1EpljWAj69i/8Os5iahICnP6m9V33gt3/8PitLpUrAqymmBsgkBGSqCFKFzPHP6qtagPj+sKKA6Jg64H8rsaEFlKmeffrLbysK6B8jQKWxFiBLf5XeFNAzVQ5YnwkoTJG/pIC+OasDqoQWST/blUqYA/qnFiCS+KwOYCX8GCoB376fLltBn/4qiicKIN1QABgKAEMBYCgADAWAoQAwFACGAsBQABgKAEMBYCgADAWAoQAwFACGAsBQABgKAEMBYCgADAWAoQAwFACGAsBQABgKAEMBYCgADAWAoQAwkQVkRIMSEDe68UIBYCgADAWAoQAwFACGAsBQABgKAEMBYCgADAWAoQAwFACGAsBQABgKAEMBYCgADAWA8RCgVx2vFmDu97xTx1VAvajgcd3wPoq45506jgIO352n+eVeP+edOg+uA9yHw0wdVsJgfASU9cBstwl7zwQFaHxaQU9v+Wx3WAUZoACNu4DjennKxVLLAW0gCqhwF3BYvUoBBQXEwD8HbIJePkcBGu86IGdPOAqeraDg151RgIb9ADCDEvDxEfckY8BDgOgB5FngC2ApQOMhoGz+7Bfz0yboVSssgjR+/YCibAKxHxAFPwEb8UZACoiBTxE0P6zEvSAWQTHwqoTLTsBxHfa2rcvzplj7aobRDKWAiAEh0Y0Xn0pYP1Rsq4TNixobw1CAxjsHtLZCc/Ne9qLpBe0UoPEvglo6YuKOtSZv6DBTgMZfQEsWEF2FjlAUoPEX0NIRYw6wx1tAa0eselpjWQewGeoQ0LSCWu+GdgaiAA37AWA4NBGMj4D9gs+Eo+E1P0BUrBwVEQe/cUGCpgbm6exuRfPNCArQ+D2QEbTdiziuu57WUICmlxwgAnU8LqAATV91QNFRRVTRJdwFkMBbQRQQOyAkuvFCAWD8Hsp3PhGLd96p49MKChuU6HjeqePfD4h0XlbCrgHPnrbEOC8FOAfcv4S1QN3Oe8PEjPkIWEAr4eQFhA5KdDzv1IFXwqkDr4RTZ3yV8MToaWyofXQTq1Odgd8LogCngIdV0O/+znkpwDGgKIG6JKilbAqr4emVgO32dPNdAngVQV0SpAD5yLKxzUoBmoCxoW2jo5cm6TsH514ldkppL+mlEhYC9gsp4GrsxO3IuI+PizSngO6AndOPnHLAdhsjzcfrzWeecGdDSHUV5idTHbdFtz0r+/1JSYDdan1yMnHz9IDYOWC8wG/GnQmIkhfGBvx29FkRRAF2AbsGvblFZ1Jdf/YuYWgFHvxm3PYy5SkgOKBjdErANtm6mALAwEfGbWXab9XmKb2a2KcjNtvlc3OrIfi8tYCtzAypNYr8OmKFWLw7ztLF1e9fCzDfUkBTQNER23/zVf6LcF4jQFQBtYE0El/gNyri8PktngCV7KWGbULpXuFRB4hVOjbLeEXQ9rYLcC1iwmJ8mqGbuWgJhT0cvhRwXfNSQEBAx+i2xsD9pJ88QxNQf31vs4kRaxtER0xvXHUC7tQNd/fvfzUW4FOUqhqgUcA10yqrBvBA5tyAPZahB+8JPjr6XEAPqTVBAZFHR5/3A65KoPCzjICepijZr5x7IcB8dS5g4i76eSbssHJu1QytudylgKuAFpWwzbqh/yUKZwEWlXDzyrnVyET0//ZwcBZgUQm7rJx7WwKlRT+jIhxWzt3eqQRSoqd7QfYr5969E5QQA7kZFzfyMTGEgVlJK/DOAWHvcaMAg38RFOlVhmknf4iASK8yZCXsGzDSqwwpwDNgrFcZpl0AhbSCIr1PmAJiB3SMjgJiB3SMDiBgUMZ7mSfsEh0FOAfsnifsEl3STaBTf/OEraOjANeAPQxL0RsRoh0fQ5gnTAFuAQMnJ11Fl2i6VzgKsHg/kuN5KSB2wJPbkmUUEDvgyW3JssQ7ws5F0HfnP+nLvbPvnZYsowCXgGUloBtBZYe4qYBpXrLs9rwU4Byw0JVwc1soYNG+5OipDrBfsowCYgdUNCxZdrtqYuJ1MPsBaLzqANnK736XYdsw3siPF8aL/72g7pnyFGCB/93Q7mEpFGCB/yS9TZwcwFaQc0C1amIeqQ6gAPeAsikUOFu7ji5xA/hRERQQOaBjdOwHxA4IiW68wHvCrANiB3SNjgIiB3SNjgKcA8rFIrq7AXbRJZ7+/jfj8rBF7ClA438rItKylRTgGtD+ZpxVdB+JG4DngNQNwOsACnAPGLUVlHolAO8HJJ4B3G9FvIqV0yOelwKcAh5Wy+gCIsQ2YlyLoDz2zTgKcA3IHBCVB1fCtyPjUjfgLCA3w3IjLdZBAW4BxWAItU5HLAGJ4yhA3YeQK9hHE5B2FnDvB8hP8TI3CoiBVw44iQXLWARFwbkO0Mke+holCtB4tIJUIXRcU0AM4DfjUocCwPg8kqx6swGPZChA4z9FKXsNWTmIAjRBU5QCHgtTgCZoVETAwAgK0ASNigjoi1GAJqQOCFk7iwI03lOUxFudA7piFKDppR/gsmBT6vQnwHLBptTpZWCWy3I1qdPL0ESXBZtSp5fBucwB9vQyPN1lwabU6Wt4esOCTe7nnTr44emJgx+enjh8IAMGPzQxcXrqCXeOoaYAjfPALKvh6d1DJihA01MR1HmnmgI0fdUBRUcrlQI0bAWBoQAwfQrguqEWUAAYCgBDAWAoAAxbQWAoAAwFgKEAMBQAhgLAUAAYCgBDAWAoAAwFgKEAMDABRAMS4BHxCAL2mZoUADlz0J9EjngEASkAHJACwAEpAByQAsABKQAckALAAUcpgNhBAWAoAAwFgKEAMBQAhgLAUAAYCgBDAWAoAAwFgKEAMP0IKO6u/HRcZ5la7Kw63rJhhVhkah4vxv03X20jc7vOZnoRUIh15W4u77guv8pFelXHWzbsTjTbqffPxYnxsJKrwdhE5nadLfQhQE2j3FwvrKKWHM2fv1bHWzYcThQtxkKt1WMTmdt1ttGHgCql7x0sfzbV8ZYNuxO9vF2dMSTGIlvKNTttInO7zjZ6ESDTpWEB0k159eZ4y4bViYrnLytZq8SKUQmwiMztOtvoQ4AqG++XkGJxzOp4y4bViXJRZIhFpGLFKFPUJjK362zjwQIKUwfHEfCkf4cUcEFz/lSLw0bL2qoMLstjFkEXNNZQueoFRKvcVAKUiRErxolUwk1ttPp9lZGad2rB8SJew7aYRjO0oZeyXyyvj4d2cMRK4zItIsVYTKMjJt9Qf3txuZo6Ig5Ux1s2rCjMzY04MepC3SYyt+tshjfjwFAAGAoAQwFgKAAMBYChADAUAIYCwFAAGAoAQwFgKAAMBYChADAUAIYCwFAAGAoAQwFgKAAMBYChADAUAIYCwFAAGAoAMzwBm6z5TYn9oAZZgxicgP3LDxFG3Y+HwQnIn/+1AP4gH87QBBzXczHrTk9TKeTUbjkC/fD5ZzElb1PuST/lxtPPeqh7XWjpHTF15bBa7l9+WZz/sQkq5teLWPSnnuehz7P640qf4iEMTYBM8vKftHDazHZiR0ztOKxmOzUnRU7N25TpWWQ6cDX1w+yISRv5bFcm8GsZ6lX/sTkq5ZZfm08pQAWbl/90pA9iaAI2cvGB+kcst8XECblx+PymMofKIJunt+q4+ON6p3j+8tmIKU2oPzZHzdw68ykEqDlHhY5w/7hScGACVJJtZOK/1vNBywSpmiqFKCCquUTV8dPJTBuVO5vsvBxTv/E6KuPLfBqj5jwPbBcNTICexiTK4LK0KbNDYb5QaVIel5V0Xgmo/0AV43qnkGX8i57NqwsZc1Qu2/J6Mp+1gHIrbQGq5NcLoDx/UZlAHZJpIpNJ/OcqB2jqneP6xzLElYCLgn2j9zbqKHOAxKSRqAQPqx91SSRRaShnsmcmvXKdtIZ6J5/9Z700BZqoA25S1ezpH/1ZHZCyAFEFC2TKyWJct3l0woh0P6xEa/GiFVT9nM8Di/nsCzGZtGpomqNmnqn5vGoFJSygbnyo9JV7olqosoKcHCqTu+wHPP9TZAh9XKF3NnL+8Hy/+GkhC3udoCZooWfL6s+rfkDCAlzpWqvhga1Jb0YqQCatath3hRo4IxWg2pRdyUsBpBMKAEMBYCgADAWAoQAwFACGAsBQABgKAEMBYCgADAWAoQAwFACGAsBQABgKAEMBYCgADAWA+T8Nxr3WlKcL5QAAAABJRU5ErkJggg==" /><!-- --></p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" title="1"><span class="co"># Generate Volcano-plot</span></a>
<a class="sourceLine" id="cb16-2" title="2"><span class="kw">ortholog_expression_plot</span>(</a>
<a class="sourceLine" id="cb16-3" title="3">  hs2mm.deg, <span class="st">&quot;volcano&quot;</span>,</a>
<a class="sourceLine" id="cb16-4" title="4">  <span class="dt">main =</span> <span class="st">&quot;Brain&quot;</span>, <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">18</span>,<span class="dv">18</span>), <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">3</span>)</a>
<a class="sourceLine" id="cb16-5" title="5">  )</a></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAYAAAAGACAMAAACTGUWNAAAApVBMVEUAAAAAACsAAFUAK1UAK4AAVYAAVaorAAArACsrAFUrKwArKysrK1UrK4ArVYArgNRVAABVACtVAFVVKwBVgIBVqqpVqtRVqv+AKwCAKyuAVSuAgFWAgKqAqoCAqtSA1NSA1P+qVQCqVSuqgFWqqlWq1ICq1P+q//++vr7UgCvUqoDU1KrU1NTU1P/U////AAD/qlX/wMv/1ID/1Kr//6r//9T////GrTAQAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAOAklEQVR4nO2dC3/rphmHlXOyuN3axV3XJt3W1d4lJ2q3RPFsff+PNnHRzRYgIV5eJP2f36/pcYwB84SLEKCsBKxk3BnYOhDADAQwAwHMQAAzEMAMBDADAcxAADMQwAwEMAMBzEAAMxDADAQwAwHMQAAzEMAMBDADAcxAADMQwAwEMAMBzEAAMxDADAQwAwHMQAAzEMAMBDADAcwsWUCeKe7++DL4/uU5u3+LnKfJrEFAln36MvQ+BBDTCsgeufPizbIFqL/80y57EK/ufv4++/RS/vp9ZeSbQ10Dqt//8o8s+3zgzu4waxDw607UAFUf7t8K3TEcGgHNL1Jk2QJqRFNfvbp/Kf9Tlfrv3kSleGwFfH4RP9NsplYh4O5H9epJv/HbP3+fVa1SI+CpbqYSZBUCMtUEqUbm8hf1q1aA+P15DwHBqfuA/+3FP7SAqtSzz3/7bQ8B9NQCVBlrAbL1V+UNAcQ0NeC5I6Com/xHCKCm0wc0BS2K/v6tUoIaQE8rQBRxpw9AJxyHRsA3L2V/FPT576J5ggDgBgKYgQBmIIAZCGAGApiBAGYggBkIYAYCmIEAZiCAGQhgBgKYgQBmIIAZCGAGApiBAGYggBkIYAYCmIEAZiCAGQhgBgKYgQBmIIAZCGBmmoB67ffwznTgwSQBeb3Vs0h0z+cCmSLg8twUe244hCFLgI8P7hwIKASc9/VG3LIwNEIpdCkfH9w5KCcUROgaMCE6KtYroOoDdBUw9gEQoKERIHZaSYzH8ECAhkhA7Oi8gIC5vL/P+vgmBOSUoyAImJ7c5OGvixkStiggfHQQQJwuIasXUFjOX5spYGbrr1izgGOWPZ6+fetOSsyJ7gYIsHOsLsCO8q8fUxEO6CbjTl/JY5EwGWeHSICYAbr8t0QNcEI0GVf/3SsVc6MjYsUCylwNf4rM0AdDQA2uA5iBAGYggBkIYAYCmIEAZiCAGQhgBgKYgQBmFi3g/f1d/ucJBMwFAiKkSwgEMAMBzEAAMxAQBu9eeIaAIKsyJBAQN80bViDAGzRBzEAAMxAwi6tm2KNV3qKAgPsDuiUupyQmx7BFAUTR+Q1KIIAZCGAGApiBAGYgICjTO2IICAoE0KRLCAQwAwHMLFVAoScTTJtfwqZLyCIFtI9cvzzPORMRAjQTBZz/1C3z/iuadAlZooD46RICAcwsVkDVD9y/HeedyAoBGp9R0N0hv38zbsEOnC4hCxUgDgUVBxCYDiMLnC4hCxUgTuIQAkzHcAROl5CFCqhrwNF4JmjQdAlZqADdB+S4Eg6C5yjIchiZxPn8AAjQ0FwHuJ8fAAEaEgE4PX08PqOgzNG84PkBE/CuAbZRKGrAePyboOODOTCeHzAafwHWCzE8P2As/gIwFREEbwHnvaUJCpguIQsV4GxdepA+P2AmCxUwL7ngzw+YwRYFUEXnBQQws0QBzWWw9UpYXAjodUPoAxwQTcbdHfQwCQIcEE7GXZ4tNy4hQOMhoG6G3JNxx/s3CHDgIUAU60N52pnviDWTcccHCHDgcyH2WBaicbFcidXFXlUWCLDjtyri9PUX+Z+Rejr08gwBdvxWRZy/O9gFhEuXkIUKkO3L8dHaBAVMl5ClChC3YsyNe+h0CVmsgKjpEgIBzCxUwMw7MVPTJWRIgH23cbij4hp8LsSc6+JCpkvIKAG9X6QhoNQbJbdwS/J9xtHU4/DuA4yXWIHTJWSUAOpMoAYw43MlvOY+IDoYBTHj1wR1Fn9Sp0sIBMxifv8IAbOAgDjpEgIBzCxYQMR0CVmsALksYtbtmIUJoLwe9tsnXDZPjSdPl5CFCrg8N4t+YqRLyEKbIHFDXoCzIoLgXwO8bspjf8A1Pn2AXPODsyLCMGOjtmOFeqB0CVmogMjpEgIBzCxRQOrnhk4asS9RQOon565fQJn22dGbEBA33WlMeqAYBJAQQAD9YpSWyQKcp8EFTpeQawG63JvFWPIf1CuzpgpwnwYXOF1v3K3RrYB3/cH64+kJGHEWVuB0vZku4PbxlK/l62v4nPWYeh3gPg0ucLqEODvh1/QELKgGlM7CW+QoyH0aXOB0Z7BKARPPa5qfLiHLFBA7XUIggJkxnTA5EGABAmhZYhMU5n7klHQJWaKAsVvD8PyAsXhMxo3YIMP//IAxzfcyBehVKVYSOD3dT0CMXvcKkk54sc8P6AiI5YJEQAI1YAzWJijGRJyAZhi6jucHJCqgMxI1dwaJPD/AXoQ3Aq6DJ3ohpv6qq754zuMkIUDjszpajUPFw9z8p0SX0ARFwacJUi1PNb4ZcVMM54Y68KkBquHJzcfiWmJJfX9A2+q8vibaCbd9wKiL4rnpErJUAWqII86uTG9t6CSMTZAq+ziXYmufjlalOFyWEBABLwEx8RFw2jmObBpx14BGwE1BWxvyawHdoJF6gBmdsHWTnvuuAQRoZg1DzaGdA6SkO+GYzLoQswR33TWAAA1RDQiXLiE2AdFuzdD0AQHTJWSpAtyjoJDpEmK5DoiXibVfB9hwXIjFAQL66JKHgBjYJ+MigZVxV8Q2sP4aYC5Rs4CInYCfgGUcW/mqC9P0vnkYCgFBmCOAJENDrFmAi4VORQgWJsDwB21Ymhi3G4aAm4BLEBAx3QCYSnRgYZaGPEstENALCAHz6C4qaV+YitTQCaMJ8gcCwqcbgIkC4rIVAUMKlIBerbn6RwQgAALImdYExR0EbUaAuQYMhKbOUJctCCgNpdoXUIeIfEMgsgCu/QGDi22vBcTfJFxupgYIHAKs63jp2JCAGwx3xNAHEDBYqIOzoRBAgluALvvoPcFmBLhGQWryCAKIGHsdAAFEDI5wICAmr7fHn5huyMTLVLkxAVdAADO4HxCTMTWAAwi4ChUlMx22ImCoaR/Yphp/Rm4TAgznv93MhkIAEXXZXu1+6d6S5FgSJIEACIjDQOnWAnrEztemBFz9CgKiclvCQkA7C81R+uVGBDR9wGtvRqgRwHMnQLEZAQMtTWcYahioRoBIQHrPD+iW/ZWAugrEzpKERgD/8wMM9Mq5FcDXAtEISPb09H5B9wWwZIhIQLLPD9iKgGRrQNl0AuKFHgXxtkFUfUCizw/o9cRKwOsaBaTy/IAbOkWtBPBNAmk2cR0g6LU1Ct0HNI0SroQp6TY1XQG9OhE7UyW9gKSeH9CbdxACbq+Oo7OR/QFdNi2AKjovsCqCmTULSG8yboAVC0h2Mq7HegWkOxXRY70Ckp2M67NeAagB42GbjEuAjw/uHAhIBLgn42Yn5g6ynBCjiNpmQABZNMESS6XwIGAFIUYBAXQhRgEBdCFGAQF0IUYBAXQhRgEBdCFGkcLcwaaBAGYggBkIYAYCmIEAZiCAGQhgBgKYgQBmIIAZCGAGApiJJ+D0tVjLpdZVPPhH44qgyLK7w5wIdEZtEQX6KpJoAs57uZju9JW1dNw4IiiqQiusBlw50Bm1RBTqq0hiCSj0kmrTmsbxEVkjUIv3jrY/S0cOdEYtEQX7KpJIAorsUeU3n1tl7RGcdmLtnmkD1YgI6oyaIwr3VSTx+gCV6+MfqnZzzjNb7RGoZsH6t+nMgRJgiyjQV5FEFnDei1WNR/9sOyJQrbatE3DnQGbUGlGYr6KIXQOu/zk7rqs3nALcORgpwB3RKFgEqAZ2BsYIRjRBzhyMbYKcEY2CXEBeN5TdXHsN4PK2yTVGMKITdubA0QmXAb5Kh8g1QH2vGfXWEYF7GOrOQeEYhob6KorooyDxleb0XI4I3BdizhwUrguxUF9FEr0POFYNyaxm0xFB7pqKcOZAZ9QSUaivIsBkHDMQwAwEMAMBzEAAMxDADAQwAwHMQAAzEMAMBDADAcxAADMQwAwEMAMBzEAAMxDADAQwAwHMQAAzEMAMBDADAcxAADMQwAwEMJOAgM5jCTrIp6WIla95Zx9QE1Qum708Pw1/1szp24nHjp+/C7EV0kKqAi7PVRHn2UP17uNAUCkgfzDIM1NM3lhXeJ4UP5ZUBTQbJHrv9gSIF1MFHCevZq6q2dSPTCIVAUW9/eWYZXd/rdeFF3c/7+SuXP22DJrrEGKR+Hn/551YIy7fqH6cdj/sq5CnnVo4rheQn/fit9U/Lj8drkJ0IxY/xO/lG81+HNPDQgKRiIBCFmLVPhxladcCjroG1G+LF00IsUHivBct1d2hFSBbriqI2F0kQuSyrqhQogvoh+hGrD6vU5Ohd8KAe8ffLNIQoDYEVV9VtTxH/Z3Fk1JE0TRv10UkQsjOQf6oftUWoHytfyk6UBG+DiW6gF6IbsTy8/WOI9XzdLYikZGGAPUlq5/t9iDxv+yhbhf0200RdcpM/uj9Bbc/ZBPTNlCiC+iF6Eas41EGVA7k271RQHhSElD9I+8IUE8KagWocm5C6NKxCaja8U//3rUCqi5gSEDn82r0+yS7hqzuQDYjoF8D9GP7ZtSA5kXzbnUVYK8BMkNV+9Zp+Dch4KoPyGU/+VS/O9AH5E0fUAt47Jd5Y7PoNEHiKqAXohPxY9va10IVm+gDbkdBagRSNoPUdrAi7HRGQbpxeb5/q1qP2xpw3uuOXPw4Pl0JaCKuPy//8gs9ZlKDgU2MgvrXAZ/+9elLrtpg3dJYrgN00yGOr/phoA+4Oxyb1urHnw7XApp09efla1niIgPqVJrVXwfcMm77+dSrYD+oU0lMwKSBX5gTkxxsYC6oR5GN335OPU0j2MJs6LaBAGYggBkIYAYCmIEAZiCAGQhgBgKYgQBmIIAZCGAGApiBAGYggBkIYAYCmIEAZiCAGQhgBgKY+T+mrQKYCObWtAAAAABJRU5ErkJggg==" /><!-- --></p>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
